### PR TITLE
[Backport v2.7-branch] kernel: mutex: fix races when lock timeout

### PR DESCRIPTION
**Pull request by zephyrbot**
Backport 89c4a074dc6257a4f2f186dbb30368937259d618 from #47906

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/48056

### CODE REVIEW

**Comment by cfriedt** [comment]
Would have been a bit better if the test described in https://github.com/zephyrproject-rtos/zephyr/issues/48056 had also been committed. I'll give this a whirl on LTS as well, just to have a bit of certainty.

**Comment by mmahadevan108** [comment]
@cfriedt Please add a test for this issue.

**Comment by cfriedt** [comment]
@mmahadevan108 - just need to get #49137 merged so I can cherry-pick the test here. It will require slight modification though for the older ztest API.

**Comment by cfriedt** [comment]
Waiting on #50958 as well now - once that is merged I'll have 2 commits to cherry pick 😀

**Comment by cfriedt** [comment]
@andyross - another ack for you

**Comment by stephanosio** [comment]
The CI issue has been fixed by #51868.

**Comment by cfriedt** [comment]
> @cfriedt Looks like there are some real failures.

I think it was the difference between a user unit test and a unit test.

**Comment by cfriedt** [comment]
@andyross @peter-mitsis - an ack from either of you here would be sufficient, I think.

**Review by cfriedt — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Pull request merged**
